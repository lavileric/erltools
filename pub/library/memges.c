/**************************************************************
     
     
                    MemGes : Malloc with a cache in files
     
                    Developped by Eric Lavillonniere on Tandon
     
   ***************************************************************/ 
#include <fcntl.h>
#include <io.h>
#include <stdio.h>
#include <sys\types.h>
#include <sys\stat.h>
#include <malloc.h>
#include "token.h"
#include "memges.h"

#define BUFFER_SIZE 4096     /* size of each cache */
#define BUFFER_NUMBER 15    /* number of caches */
/* the name of temporary file */ 
/* #define TMP_NAME "d:\cache.fil" */ 

extern char *TMP_NAME ;

typedef struct free_elem {
    long    deb ;
    long    size ;
    struct free_elem *Next ;
}   FREE_ELEM,  *PFREE_ELEM ;

static long theTime ;
static long maxpos = 0 ;
static long posBuffer [BUFFER_NUMBER];                  /* the beginning */ 
                                                        /* position of */ 
                                                        /* each cache */ 
static long timeBuffer [BUFFER_NUMBER];                 /* time of last */ 
                                                        /* use */ 
static char writeBuffer [BUFFER_NUMBER];
static char contBuffer [BUFFER_NUMBER][BUFFER_SIZE];    /* the cache */ 
static int  currentBuffer = 0 ;
static int  cacheFile ;                                 /* the file used */ 
                                                        /* by the cache */ 
                                                        /* manager */ 
static PFREE_ELEM   freeList = (PFREE_ELEM)0 ;
static PFREE_ELEM   keepElem = (PFREE_ELEM)0 ;

/**************************************************************
       CacheRead : read a value with a cache
   ***************************************************************/ 
CacheRead ( position )
long    position ;
{
    int cache = 0 ;
    long    cachePosition ;
    int *ret ;
    
    for (; cache < BUFFER_NUMBER ; cache++ ) {
    deb : 
        cachePosition = posBuffer [cache];
        if ( position >= cachePosition
                 && position <= cachePosition + BUFFER_SIZE - 1 ) {
            ret = (int *)&contBuffer [cache][position - cachePosition];
            timeBuffer [cache] = theTime++ ;
            return *ret ;
        }
    }
    BufferPosition(position);/* we don't find a cache so we put one */ 
    cache = currentBuffer ;
    goto deb ;
}

/**************************************************************
       CacheAddrRead : read an address of a value in a cache
   ***************************************************************/ 
char *CacheAddrRead ( position )
long    position ;
{
    int cache = 0 ;
    long    cachePosition ;
    
    for ( cache = 0 ; cache < BUFFER_NUMBER ; cache++ ) {
    deb : 
        cachePosition = posBuffer [cache];
        if ( position >= cachePosition
                 && position <= cachePosition + BUFFER_SIZE - 1 ) {
            timeBuffer [cache] = theTime++ ;
            return &contBuffer [cache][position - cachePosition];
        }
    }
    BufferPosition(position);/* we don't find a cache so we put one */ 
    cache = currentBuffer ;
    goto deb ;
}

/**************************************************************
       CacheWrite : write a value with a cache
   ***************************************************************/ 
CacheWrite ( position, value )
long    position ;
int value ;
{
    int cache, *pos ;
    long    cachePosition ;
    
    for ( cache = 0 ; cache < BUFFER_NUMBER ; cache++ ) {
    deb : 
        cachePosition = posBuffer [cache];
        if ( position >= cachePosition
                 && position <= cachePosition + BUFFER_SIZE - 1 ) {
            pos = (int *)&contBuffer [cache][position - cachePosition];
            timeBuffer [cache] = theTime++ ;
            writeBuffer [cache] = 1 ;
            *pos = value ;
            return ;
        }
    }
    BufferPosition(position);/* we don't find a cache so we put one */ 
    cache = currentBuffer ;
    goto deb ;
}

/**************************************************************
       LCacheRead : read a long value with a cache
   ***************************************************************/ 
long LCacheRead ( position )
long    position ;
{
    int cache = 0 ;
    long    cachePosition ;
    long    *pos ;
    
    for ( cache = 0 ; cache < BUFFER_NUMBER ; cache++ ) {
    deb : 
        cachePosition = posBuffer [cache];
        if ( position >= cachePosition
                 && position <= cachePosition + BUFFER_SIZE - 1 ) {
            timeBuffer [cache] = theTime++ ;
            pos = (long *)&contBuffer [cache][position - cachePosition];
            return *pos ;
        }
    }
    BufferPosition(position);/* we don't find a cache so we put one */ 
    cache = currentBuffer ;
    goto deb ;
}

/**************************************************************
       LCacheWrite : write a value with a cache
   ***************************************************************/ 
LCacheWrite ( position, value )
long    position ;
long    value ;
{
    int cache ;
    long    cachePosition, *pos ;
    
    for ( cache = 0 ; cache < BUFFER_NUMBER ; cache++ ) {
    deb : 
        cachePosition = posBuffer [cache];
        if ( position >= cachePosition
                 && position <= cachePosition + BUFFER_SIZE - 1 ) {
            pos = (long *)&contBuffer [cache][position - cachePosition];
            timeBuffer [cache] = theTime++ ;
            writeBuffer [cache] = 1 ;
            *pos = value ;
            return ;
        }
    }
    BufferPosition(position);/* we don't find a cache so we put one */ 
    cache = currentBuffer ;
    goto deb ;
}

/**************************************************************
       BufferPosition : put a cache to a position
   ***************************************************************/ 
BufferPosition ( position )
long    position ;
{
    int i ;
    long    min = timeBuffer [0];
    
    currentBuffer = 0 ;
    for ( i = 1 ; i < BUFFER_NUMBER ; i++ ) 
        if ( timeBuffer [i] < min ) {
            min = timeBuffer [i];
            currentBuffer = i ;
        }
    if ( writeBuffer [currentBuffer] ) {
        
        /* save current data */ 
        if ( posBuffer [currentBuffer] >= 0
                 && posBuffer [currentBuffer] < maxpos ) {
            lseek(cacheFile, posBuffer [currentBuffer], SEEK_SET);
            if ( write(cacheFile, (char *)contBuffer [currentBuffer]
                    , BUFFER_SIZE)
                 == - 1 ) {
                printf("File system is full \n");
                exit( - 1);
            }
        } else if ( posBuffer [currentBuffer] >= maxpos ) {
            write(2, " Illegal access to file system \n", 32);
            exit(0);
        }
    }
    
    /* read the new data */ 
    if ( (posBuffer [currentBuffer]
             = (long)position / BUFFER_SIZE * BUFFER_SIZE)
         > maxpos ) {
        write(2, " Illegal read for  file system \n", 32);
        exit(0);
    }
    lseek(cacheFile, posBuffer [currentBuffer], SEEK_SET);
    if ( posBuffer [currentBuffer] == maxpos ) {
        
        /*        write(cacheFile,contBuffer[0],BUFFER_SIZE);  */ 
        writeBuffer [currentBuffer] = 0 ;
        maxpos += BUFFER_SIZE ;
    } else {
        read(cacheFile, (char *)contBuffer [currentBuffer], BUFFER_SIZE);
        writeBuffer [currentBuffer] = 0 ;
    }
}

/**************************************************************
       FreeElemAlloc : alloc a free list elem
   ***************************************************************/ 
PFREE_ELEM FreeElemAlloc ()
{
    PFREE_ELEM  ret ;
    
    if ( keepElem ) {
        ret = keepElem ;
        keepElem = keepElem -> Next ;
    } else 
        ret = (PFREE_ELEM)malloc(sizeof(FREE_ELEM));
    return ret ;
}

/**************************************************************
       CacheInit : init the cache manager
   ***************************************************************/ 
CacheInit ()
{
    int i ;
    
    cacheFile = open(TMP_NAME, O_CREAT | O_RDWR | O_BINARY
        , S_IREAD | S_IWRITE);
    for ( i = 0 ; i < BUFFER_NUMBER ; i++ ) {
        posBuffer [i] = (long) - (BUFFER_SIZE + 1);
        timeBuffer [i] = 0 ;
        writeBuffer [i] = 0 ;
    }
    theTime = 0 ;
    freeList = FreeElemAlloc();
    freeList -> deb = (long)0 ;
    freeList -> size = (long)10000000 ;
    freeList -> Next = (PFREE_ELEM)0 ;
    CacheMalloc(1);/* suppress null pointer */ 
}

/**************************************************************
       CacheEnd : end the cache manager
   ***************************************************************/ 
CacheEnd ()
{
    close(cacheFile);
    remove(TMP_NAME);
}

/**************************************************************
       CacheMalloc : malloc of a data with cache manager
   ***************************************************************/ 
long CacheMalloc ( size )
int size ;
{
    PFREE_ELEM  ptFreeElem = freeList ;
    PFREE_ELEM  ptOld = freeList ;
    int exit = 0 ;
    long    ret ;
    
    size += sizeof(int);
    while ( ptFreeElem && ptFreeElem -> size < size ) {
        ptOld = ptFreeElem ;
        ptFreeElem = ptFreeElem -> Next ;
    }
    if ( !ptFreeElem ) 
        return (long)0 ;
    ret = ptFreeElem -> deb ;
    if ( (long)ret / BUFFER_SIZE * BUFFER_SIZE + BUFFER_SIZE < ret + size ) 
        exit = 1 ;
    else {
        if ( ptFreeElem -> size < size + 4 * sizeof(int) ) 
            size = ptFreeElem -> size ;/* do not tolerate unusable free elem */ 
        CacheWrite(ret, size);
    }
    ret += sizeof(int);
    if ( ptFreeElem -> size != size ) {
        ptFreeElem -> deb += size ;
        ptFreeElem -> size -= size ;
    } else {
        if ( ptOld == ptFreeElem ) {
            freeList = freeList -> Next ;
            ptOld -> Next = keepElem ;
            keepElem = ptOld ;
        } else {
            ptOld -> Next = ptFreeElem -> Next ;
            ptFreeElem -> Next = keepElem ;
            keepElem = ptFreeElem ;
        }
    }
    if ( !exit ) 
        return ret ;
    return CacheMalloc(size);
}

/**************************************************************
       CacheFree : free of a data with cache manager
   ***************************************************************/ 
CacheFree ( position )
long    position ;
{
    PFREE_ELEM  ptFreeElem = freeList, ptOld = freeList, ptNew ;
    int size, bind = 0 ;
    
    position -= sizeof(int);
    while ( ptFreeElem && ptFreeElem -> deb <= position ) {
        ptOld = ptFreeElem ;
        ptFreeElem = ptFreeElem -> Next ;
    }
    size = CacheRead(position);
    if ( ptOld != ptFreeElem && position < ptOld -> deb + ptOld -> size
             || position + size > ptFreeElem -> deb ) {
        write(2, "double free \n", 13);
        return ;
    }
    if ( ptFreeElem ) 
        if ( position + size == ptFreeElem -> deb ) {
            ptFreeElem -> deb -= size ;
            ptFreeElem -> size += size ;
            bind = 1 ;
        }
    if ( ptOld -> deb + ptOld -> size == position ) {
        if ( bind ) {
            ptOld -> size += ptFreeElem -> size ;
            ptOld -> Next = ptFreeElem -> Next ;
            ptFreeElem -> Next = keepElem ;
            keepElem = ptFreeElem ;
        } else 
            ptOld -> size += size ;
        return ;
    }
    if ( !bind ) {
        ptNew = FreeElemAlloc();
        ptNew -> size = size ;
        ptNew -> deb = position ;
        ptNew -> Next = ptFreeElem ;
        if ( ptOld != ptFreeElem ) 
            ptOld -> Next = ptNew ;
        else 
            freeList = ptNew ;
    }
}
