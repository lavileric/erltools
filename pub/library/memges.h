/**************************************************************
     
     
                    QMemGes : Malloc without a cache in files
     
                    Developped by Eric Lavillonniere on Tandon
     
   ***************************************************************/
/* This file is part of metagen                                             
   
   Metagen is a syntaxic analyser generator with backtrack.
   
   It runs on dos and unix. It generates c code. */
/* Copyright(C) 1989 Eric Lavillonniere */
/*
   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 1, or (at your option)
   any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program; see the file COPYING. If not, write to 
   the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
   
   */
#ifndef MEMGES_FILE
#   define MEMGES_FILE 1
#   include <stdio.h>
    
    /**************************************************************
           CacheRead : read a value with a cache
       ***************************************************************/
    inline int CacheRead ( void *position )
    {
        return *((int *)position);
    }
    
    /**************************************************************
           CacheAddrRead : read an address of a value in a cache
       ***************************************************************/
    inline char *CacheAddrRead ( void *position )
    {
        return (char *)position ;
    }
    
    inline char *CacheAddrReadM ( void *position )
    {
        return (char *)position ;
    }
    
    /**************************************************************
           CacheWrite : write a value with a cache
       ***************************************************************/
    inline void CacheWrite ( void *position, int value )
    {
        register int    *addr = (int *)position ;
        
        *addr = value ;
    }
    
    /**************************************************************
           LCacheRead : read a long value with a cache
       ***************************************************************/
    inline void *LCacheRead ( void *position )
    {
        return *((void **)position);
    }
    
    /**************************************************************
           LCacheWrite : write a value with a cache
       ***************************************************************/
    inline void LCacheWrite ( void **position, void *value )
    {
        *position = value ;
    }
    
    /**************************************************************
           CacheInit : init the cache manager
       ***************************************************************/
    inline void CacheInit () {}
    
    /**************************************************************
           CacheEnd : end the cache manager
       ***************************************************************/
    inline void CacheEnd () {}
    
    /**************************************************************
           CacheMalloc : malloc of a data with cache manager
       ***************************************************************/
    extern void MetaExit (int, const char *) ;
    
    inline long CacheMalloc ( int size )
    {
        long    ret ;
        
        if ( (ret = (long)malloc(size)) ) 
            return ret ;
        else {
            MetaExit(3, "Allocation Error\n");
        }
        return 0 ;
    }
    
    /**************************************************************
           CacheFree : free of a data with cache manager
       ***************************************************************/
    inline void CacheFree ( void *position )
    {
        free((char *)position);
    }
    
    inline int CacheLockFree ()
    {
        return 0 ;
    }
    
    inline void CacheUnLockFree () {}
#endif
