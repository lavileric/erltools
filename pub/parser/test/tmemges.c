#include "token.h"
#include "memges.h"
#include <stdlib.h>
char TMP_NAME[50];
#define SIZE 0xfff

typedef struct elem {
	struct elem __far * next ;
	long content ;
} ELEM , __far * PELEM ;

main ()
{
	long i ,j;
	long value ;
	char string [20];
	PELEM courant,precedent;
	strcpy (TMP_NAME ,getenv ("TMP"));
	strcat (TMP_NAME ,"cache.fil");
	CacheInit();

	precedent = (PELEM) CacheMalloc(sizeof(ELEM));
	LCacheWrite( (long) precedent ,(long) 0);
	LCacheWrite( (long) precedent + sizeof(PELEM), (long) 0);
	for (i = 0 ; i <=  SIZE; i++) {
		j =CacheMalloc(sizeof(ELEM));
		LCacheWrite(j,(long) precedent);
		LCacheWrite(j+sizeof(long),i);
		precedent = (PELEM) j;
	}

	courant = precedent ;
	while ( courant != (PELEM) 0) {
		precedent = courant ;
		courant = (PELEM) LCacheRead((long) courant) ;
		CacheFree((long) precedent);
	}

	precedent = (PELEM) CacheMalloc(sizeof(ELEM));
	LCacheWrite( (long) precedent ,(long) 0);
	LCacheWrite( (long) precedent + sizeof(PELEM), (long) 0);
	for (i = 0 ; i <=  SIZE; i++) {
		j =CacheMalloc(sizeof(ELEM));
		LCacheWrite(j,(long) precedent);
		LCacheWrite(j+sizeof(long),i);
		precedent = (PELEM) j;
	}

	courant = (PELEM) j;
	for (i = SIZE ; i ; i--) {
		courant = (PELEM) LCacheRead((long)courant);
		value = LCacheRead((long) courant + sizeof(long));
		printf ("%d ",i);
		_ltoa((long) courant,string,10);
		printf ("%s ",&string);
		_ltoa((long) courant,string,16);
		printf ("%s ",&string);
		_ltoa((long) value,string,10);
		printf ("%s ",&string);
		ltoa((long) value,string,16);
		printf ("%s \n",&string);
	}

}
