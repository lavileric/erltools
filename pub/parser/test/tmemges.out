Microsoft PLIST Version 1.20

Profile: Function timing, sorted by time.
Date:    Tue Jan 12 23:52:10 1993


Program Statistics
------------------
    Total time: 88769.187 milliseconds
    Time outside of functions: 19.408 milliseconds
    Call depth: 5
    Total functions: 13
    Total hits: 53293
    Function coverage:  76.9%

Module Statistics for c:\meta\parser\test\tmemges.exe
-----------------------------------------------------
    Time in module: 88749.780 milliseconds
    Percent of time in module: 100.0%
    Functions in module: 13
    Hits in module: 53293
    Module function coverage:  76.9%

         Func          Func+Child        Hit 
         Time   %         Time      %   count Function
------------------------------------------------------
    73936.928  83.3     88749.780 100.0     1 main (tmemges.c:13)
     5370.384   6.1      5370.384   6.1 16388 LCacheWrite (bmemges.c:318)
     4118.028   4.6      4118.028   4.6 12287 LCacheRead (bmemges.c:254)
     2683.863   3.0      2694.571   3.0  8195 CacheWrite (bmemges.c:177)
     1292.875   1.5      1292.875   1.5  4097 CacheRead (bmemges.c:65)
      973.318   1.1      3667.889   4.1  8205 CacheMalloc (bmemges.c:440)
      351.828   0.4      1645.096   1.9  4097 CacheFree (bmemges.c:496)
       11.416   0.0        13.889   0.0     1 CacheInit (bmemges.c:411)
       10.708   0.0        10.708   0.0    11 AllocBuffer (bmemges.c:42)
        0.432   0.0         0.432   0.0    11 FreeElemAlloc (bmemges.c:395)
        0.000   0.0         0.000   0.0     0 CacheAddrRead (bmemges.c:135)
        0.000   0.0         0.000   0.0     0 CacheAddrReadM (bmemges.c:155)
        0.000   0.0         0.000   0.0     0 CacheEnd (bmemges.c:430)
